django/db/migrations/state.py

```diff
@@ -184,14 +210,58 @@ def add_field(self, app_label, model_name, name, field, preserve_default):
         else:
             field = field
         self.models[app_label, model_name].fields[name] = field
+        if 'all_relations' in self.__dict__:
+            remote_field = field.remote_field
+            if remote_field:
+                model_key = (app_label, model_name)
+                concretes, _ = self._get_concrete_models_mapping_and_proxy_models()
+                real_apps = set(self.real_apps)
+                remote_model_key = resolve_relation(remote_field.model, *model_key)
+                if remote_model_key[0] not in real_apps and remote_model_key in concretes:
+                    remote_model_key = concretes[remote_model_key]
+                self.relations[remote_model_key][model_key].append((name, field))
+                through = getattr(remote_field, 'through', None)
+                if through:
+                    through_model_key = resolve_relation(through, *model_key)
+                    if through_model_key[0] not in real_apps and through_model_key in concretes:
+                        through_model_key = concretes[through_model_key]
+                    self.relations[through_model_key][model_key].append((name, field))
+        else:
+            self.all_relations
         # Delay rendering of relationships if it's not a relational field.
         delay = not field.is_relation
         self.reload_model(app_label, model_name, delay=delay)
 
     def remove_field(self, app_label, model_name, name):
         model_state = self.models[app_label, model_name]
         old_field = model_state.fields.pop(name)
-        # Delay rendering of relationships if it's not a relational field.
+        if 'all_relations' in self.__dict__:
+            remote_field = old_field.remote_field
```

Comment:
We can reuse `resolve_model_field_relations()` proposed in the previous comments, and switch `append` to `remove` when `field_name` is not in `model_state.fields`.